---
title: "Part 1 - HIV in Africa"
author: "Armand"
date: "08 April 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Part 1, HIV in In Sub-Saharan Africa

This 4 part series covers topics of HIV AIDS.  In this first part, we discuss the HIV pandemic in Sub-Saharan Africa.  In the coming sections, we will discuss a recent publication in [`PLoS ONE`](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0213241).  The authors described how they used affordable hardware to create a [`phylogenetic`](https://en.wikipedia.org/wiki/Phylogenetics) pipeline, tailored for the HIV drug resistance testing facility.  The third section will cover analysing inter- and intrapatient genetic distances of HIV genes using R.  In the final section, we will show how to use R to create distance matrix heat maps to visualize genetic distances.  But first, we need to understand the burden and challenges of HIV.

### Transmission

Worldwide, approximately 36.9 million (UNAIDS) people are living with HIV. 

HIV is transmitted mainly by 

- Having unprotected sex.
- Non-sterile needles in drug use or sharing needles.
- Mother to child transmission during birth or breastfeeding.
- Infected blood transfusions, transplants or other medical procedures. However, this is very unlikely.


### Treatment


HIV is treatable but not (yet) curable. The good news however, is that if a person receives Antiretroviral (ARV) treatment, their viral load suppresses and the chance of transmitting HIV, drastically decreases. So 30 years into this pandemic, the big question is, why is HIV still a problem?


Politics, stigma and stupidity aside.  Not all countries adopted the use of ARVs in an equal manner.  Although AZT (Zidovudine) was the first drug to be approved by the [`FDA`](https://www.fda.gov/forpatients/illness/hivaids/history/ucm151074.htm) in March 1987, it was soon discovered that monotherapy with only AZT was not effective for very long as the virus developed resistance to the medicine quickly.  Since then ARVs have come a long way and patients are placed on HAART (Highly Active Antiretroviral Treatment) or cART (combination Antiretroviral Treatment) which typically consists of 3 drugs of different classes.  

In South Africa, the first AIDS-related death occurred in 1985. Not all patients were eligible to receive ARVs and it was only in 2004 that ARVs became publicly available in South Africa. Eligibility restriction still applied so not all HIV infected patients received treatment. Ideally, a country would have all it’s HIV infected people on treatment, but due to financial constraints, this is not always possible. In South Africa patients where only initialized on ARVs when their CD4 counts dropped below a certain level. This threshold was initially 200 cells/mL in 2004 which was then changed to 350 cells/mL and 500 cell/mL at later intervals. These recommendations were a compromise between the availability of funds and getting ARVs to the people needing it the most. CD4 cells are a major component of the immune system; the lower the CD4 cell count the higher the chance for opportunistic infections. Thus, the idea is to support the patients who are most likely to contract an opportunistic infection. The problem with this was that about only a third of the HIV infected people in South Africa were receiving HAART treatment. In 2017 the guidelines changed to test and treat, i.e. any newly diagnosed patient will receive HAART treatment. This is a big improvement for many reasons, but notably a lower infection rate. If a patient is taking HAART treatment and it is effective in suppressing the viral replication, the chances of the patient transmitting the virus are very close to zero. However, these treatments are not without side effects which in some cases causes very poor adherence to the treatment. There are numerous factors to blame here, but socio-economic factors and depression are worth mentioning. There is also ignorance and the ‘fear of knowing’ which causes people not to know their status. Finally, the x-factor or human nature brings with it various other complexities such as conspiracy theories, religious and personal beliefs. This will be a very long post if we delve into all the issues, the take home: the situation is complicated.

This being a blog about R, let's look at some ARV statistics and see if we can make any conclusions.


```{r }

library(dplyr)
library(readr)
library(stringr)
library(tidyr)
library(ggplot2)
library(knitr)
library(kableExtra)
library(maptools)
library(viridis)
library(RColorBrewer)
```


Read in the data.

```{r}
ARV_cov.df <- read_csv("ARV cov 2017.csv", col_names = TRUE)

# A list of Sub-Saharan countries

ssa.list <- (read_csv("Sub-Saharan.txt", col_names = FALSE)[["X1"]])

head(ARV_cov.df)
```

[`Data`](http://apps.who.int/gho/data/node.main.626)

There is a couple of issues with this data.

 - Very long variable names, descriptive, but difficult to work with
 - The values contain confidance intervals in brackets, this will be difficult to work with as is
 - We might want to transform no data to NA
 - We are interested in Sub-Saharan Africa, but the data is for the whole world


Luckily for us, this is a post about R. Fixing the above issues should be a breeze.


```{r}
# rename the variables
colnames(ARV_cov.df) <- c("Country", "PersCov", "NumberOnARV", "NumberInfected")

# Remove the ranges in brackets and convert the values to intigers

ARV_cov.df <- ARV_cov.df %>% 
  na_if("No data") %>% 
  mutate(PersCov = str_replace_all(PersCov, "\\[.*?\\]", ""),
         NumberInfected = str_replace_all(NumberInfected, "\\[.*?\\]", ""),
         NumberInfected = str_replace_all(NumberInfected, "<", ""),
         NumberInfected = str_replace_all(NumberInfected, " ", ""),
         PersCov = as.double(PersCov),
         NumberOnARV = as.double(NumberOnARV),
         NumberInfected = as.double(NumberInfected))

# select only the countries we need

ARV_cov.df <- ARV_cov.df[ARV_cov.df$Country %in% ssa.list, ]

head(ARV_cov.df)
```

In the code above we renamed the column headers to Country, PersCov (percentage ARV coverage), NumberOnARv (Number of patients on ARVs) and NumberInfected (Number of patients infected).  We used a regular expression to get rid of all the ranges.  We removed the "<" sign and spaces within numbers.  Changed "No data" to NA and converted the characters to numbers. We filtered out the countries we don't want.  Note some countries are not available in the ARV data, like Swaziland and Reunion.


Let's see how these countries rank by ARV coverage

```{r, fig.width=8, fig.height=6}
ARV_cov.df %>% 
  na.omit(PersCov) %>% 
  ggplot(aes(x = reorder(Country, PersCov), y = PersCov, fill = log10(NumberInfected)))+
  scale_fill_viridis(name = "Log10 number infected")+
  geom_col()+
  coord_flip()+
  ylab("% ARV coverage") + xlab("Country")+
  theme_bw()


```


From the graph above we can see that South Africa has the highest infection rate and plenty of catchup to do.  It is great to see that Zimbabwe, a country in the limelight for various problems, has such a good ARV program.

```{r}

ARV_cov.df %>% 
  filter(NumberInfected >= max(ARV_cov.df$NumberInfected, na.rm = TRUE)) %>% 
  kable(format = "html", caption = "Country with the highest HIV infection") %>% 
  kable_styling(bootstrap_options = c("striped", "hover"))


```


### HIV related deaths

We can also look at HIV related deaths. [`Data`](http://apps.who.int/gho/data/node.main.623?lang=en)

```{r}
HIV_mort <- read_csv("HIV deaths.csv", col_names = TRUE)
head(HIV_mort)



# Remove the ranges in brackets and convert the values to intigers



HIV_mort <- HIV_mort %>% 
  na_if("No data") %>% 
  mutate(Deaths_2017 = str_replace_all(Deaths_2017, "\\[.*?\\]", ""),
         Deaths_2010 = str_replace_all(Deaths_2010, "\\[.*?\\]", ""),
         Deaths_2005 = str_replace_all(Deaths_2005, "\\[.*?\\]", ""),
         Deaths_2000 = str_replace_all(Deaths_2000, "\\[.*?\\]", ""),

         Deaths_2017 = str_replace_all(Deaths_2017, "<", ""),
         Deaths_2010 = str_replace_all(Deaths_2010, "<", ""),
         Deaths_2005 = str_replace_all(Deaths_2005, "<", ""),
         Deaths_2000 = str_replace_all(Deaths_2000, "<", ""),

         Deaths_2017 = str_replace_all(Deaths_2017, " ", ""),
         Deaths_2010 = str_replace_all(Deaths_2010, " ", ""),
         Deaths_2005 = str_replace_all(Deaths_2005, " ", ""),
         Deaths_2000 = str_replace_all(Deaths_2000, " ", ""),
  
         Deaths_2017 = as.double(Deaths_2017),
         Deaths_2010 = as.double(Deaths_2010),
         Deaths_2005 = as.double(Deaths_2005),
         Deaths_2000 = as.double(Deaths_2000))

# select only the countries we need
HIV_mort <- HIV_mort[HIV_mort$Country %in% ssa.list, ]

head(HIV_mort)

```


The above dataframe is now clean and we can do some sniffing around to get more insights.

```{r}
summary(HIV_mort)
```

From the table above we can see that the 2017 mean for the dataset as a whole is about half of that during the early 2000s.  It would be interesting to plot this data, but as it will probably be to busy as it is.  We can have a look at countries which had the most change rather.

```{r}

HIV_mort <- HIV_mort %>% 
  mutate(min = apply(HIV_mort[, 2:4], 1, FUN = min)) %>% 
  mutate(max  = apply(HIV_mort[, 2:4], 1, FUN = max)) %>% 
  mutate(Change = max - min)

```


Next we can create a plot of the data and look at the top 10 countries with the biggest change in HIV related mortality.



```{r}

HIV_mort %>% 
  top_n(10, wt = Change) %>% 
  gather(Year, Deaths, Deaths_2017:Deaths_2000) %>% 
  na.omit() %>% 
  separate(Year, c("D", "Year"), sep = "_", convert = TRUE) %>% # I should be able to use NA rather than D, but get an error
  ggplot(aes(x = Year, y = Deaths, color = Country))+
  geom_line()+
  geom_vline(xintercept = 2004, color = "black", linetype  = "dotted")+
  theme_bw()+
  theme(legend.position = "bottom")
  
```



Remember we mentioned that HAART was introduced in 2004 in South Africa, depicted here by the black dotted line.  It is easy to appreciate the dramatic effect the introduction of ARVs had in South Africa.  Although, the picture above is positive, the fight is not over.  The target is to get at least 90% of HIV infected patients on treatment.  Adherence to ARV regimens stays crucial not only to suppress viral replication, but also to minimize development of drug resistance. 


We have now looked at countries ranked by ARV coverage and we have seen the impact of the introduction of ARVs.  


We can also look at the rates of HIV infection in different African countries.  


The world fact book  by the CIA has some HIV infection rate data.

[`Data`](https://www.cia.gov/LIBRARY/publications/the-world-factbook/rankorder/rawdata_2155.txt)



```{r}
# load the HIV data
HIV_rate_2016 <- read_csv("HIV rates.csv", col_names = TRUE)
#read in the shape file 
africa <- readShapeSpatial("Africa_SHP/Africa.shp") %>% 
  fortify(africa.shp, region = "COUNTRY") %>% 
  merge(HIV_rate_2016, by.x = "id", by.y = "Country", all.x=TRUE) %>% 
  arrange(order)

africa %>%   
  ggplot()+
    geom_polygon(aes(x = long, y = lat, group = group, fill = Rate),
                 color = "black", size = 0.2)+
    coord_map()+
    scale_fill_distiller(palette = "Greens", direction = 1, name = "% HIV infection rate") +
    theme(panel.background = element_blank(), axis.text = element_blank(),
          axis.title = element_blank(), axis.ticks = element_blank())



```

As we can see from the map above, for some countries we have no data.  We have used R to sketch some ideas regarding this pandemic.  From our ARV coverage data and the plot showing the effect of ARV on HIV related deaths we can see that change is happening.  However, the choropleth above shows that we still have a long way to go to control this outbreak.





In the next section we will talk more about drug resistance and how sequencing viral genes reveal which drugs a patient will still respond to.



























